src/pyhf/pdf.py

```diff
@@ -464,6 +464,12 @@ def set_poi(self, name):
             raise exceptions.InvalidModel(
                 f"The parameter of interest '{name:s}' cannot be fit as it is not declared in the model specification."
             )
+        if dict(self.modifiers).get(name, "") == "shapefactor":
```

Comment:
since the entire thing is built up, can't you go and look at the paramset instead of in modifiers? I would've thought we have information about the modifier type in paramset... but maybe not.

Reply:
The type is not in there as far as I can see, but `self.param_set(name).n_parameters` (plus optionally `self.param_set(name).constrained`) could serve as a proxy. Having now looked at this, `shapesys` will suffer from the same issue with the assert so we can also switch to just raising for any cases of `n_parameters>1`.